### ISSUE TITLE
Is USE_NIS usable?

### ISSUE THREAD

**Issue by alejandro-colomar** [shadow-maint/shadow#909]
I don't see any code that allows enabling NIS.  Maybe I'm grepping for the wrong thing?

If there's no way to enable it, we could kill all that as dead code.

```sh
$ grep -rni use_nis
lib/prefix_flag.c:120:#ifdef	USE_NIS
lib/shadow.c:21:#ifdef	USE_NIS
lib/shadow.c:40:#ifdef	USE_NIS
lib/shadow.c:82:#ifdef	USE_NIS
lib/shadow.c:153:#ifdef	USE_NIS
lib/shadow.c:170:#ifdef	USE_NIS
lib/shadow.c:189:#ifdef	USE_NIS
lib/shadow.c:210:#ifdef	USE_NIS
lib/shadow.c:243:#ifdef	USE_NIS
lib/shadow.c:265:#ifdef	USE_NIS
lib/shadow.c:287:#ifdef	USE_NIS
lib/shadow.c:309:#ifdef	USE_NIS
lib/shadow.c:338:#ifdef	USE_NIS
lib/shadow.c:348:#ifdef	USE_NIS
lib/shadow.c:364:#ifdef	USE_NIS
lib/shadow.c:372:#ifdef	USE_NIS
lib/shadow.c:460:#ifdef	USE_NIS
lib/shadow.c:467:#ifdef	USE_NIS
lib/shadow.c:501:#ifdef	USE_NIS
lib/shadow.c:519:#ifdef	USE_NIS
lib/gshadow.c:33:#ifdef	USE_NIS
lib/gshadow.c:47:#ifdef	USE_NIS
lib/gshadow.c:93:#ifdef	USE_NIS
lib/gshadow.c:157:#ifdef	USE_NIS
lib/gshadow.c:212:#ifdef	USE_NIS
lib/gshadow.c:240:#ifdef	USE_NIS
lib/gshadow.c:256:#ifdef	USE_NIS
lib/gshadow.c:264:#ifdef	USE_NIS
lib/gshadow.c:356:#ifdef	USE_NIS
lib/gshadow.c:363:#ifdef	USE_NIS
lib/gshadow.c:399:#ifdef	USE_NIS
lib/gshadow.c:410:#ifdef	USE_NIS
src/chfn.c:674:#ifdef	USE_NIS
src/chsh.c:525:#ifdef	USE_NIS
src/groupmod.c:837:#ifdef	USE_NIS
src/groupdel.c:436:#ifdef	USE_NIS
src/userdel.c:1132:#ifdef	USE_NIS
src/userdel.c:1151:#endif				/* USE_NIS */
src/usermod.c:265:#ifdef	USE_NIS
src/usermod.c:1279:#ifdef	USE_NIS
src/useradd.c:820:#ifdef	USE_NIS
```

```c
$ grep -rnIi '\<nis\>'  | grep -v -e ^po -e ^NEWS -e ^man -e ^ChangeLog -e ^tests
contrib/adduser.c:387:      printf ("\nIs this correct? [y/N]: ");
doc/HOWTO:227:     NFS (Network File System), NIS, or some other method.
doc/HOWTO:882:  If you insist on the standard one, or you need NIS support, Sunsite
doc/HOWTO:883:  has ftpd-shadow-nis.tgz
doc/HOWTO:885:  shadow-nis.tgz>
lib/prefix_flag.c:122:		__setspNIS(0); /* disable NIS for now, at least until it is properly supporting a "prefix" */
lib/commonio.c:528: * New entries are inserted before the first NIS entry.  Order is preserved
lib/commonio.c:540: * Insert an entry between the regular entries, and the NIS entries.
lib/commonio.c:737:	struct commonio_entry *nis = NULL;
lib/commonio.c:753:		nis = ptr;
lib/commonio.c:769:	     nis != ptr;
lib/commonio.c:786:	if (NULL == nis)
lib/commonio.c:795:	entries[n]->next = nis;
lib/shadow.c:48: * __setspNIS - turn on or off NIS searches
lib/shadow.c:61: * bind_nis - bind to NIS server
lib/shadow.c:154:	 * to turn on NIS processing.
lib/shadow.c:168:	 * incorrectly formatted number, unless we are using NIS.
lib/shadow.c:371:		 * If this entry began with a NIS escape character, we have
lib/shadow.c:384:		 * If this isn't a NIS user and this isn't an escape to go
lib/shadow.c:385:		 * use a NIS map, it must be a regular local user.
lib/shadow.c:392:		 * If this is an escape to use an NIS map, switch over to
lib/shadow.c:485:	 * other three files.  I can't just blindly turn off NIS because
lib/shadow.c:487:	 * that case, I never discover that NIS is present.
lib/gshadow.c:49: * bind_nis - bind to NIS server
lib/gshadow.c:283:		 * If this entry began with a NIS escape character, we have
lib/gshadow.c:297:		 * If this isn't a NIS group and this isn't an escape to go
lib/gshadow.c:298:		 * use a NIS map, it must be a regular local group.
lib/gshadow.c:306:		 * If this is an escape to use an NIS map, switch over to
etc/login.access:30:# If you run NIS you can use @netgroupname in host or user patterns; this
src/chfn.c:683:		         _("%s: cannot change user '%s' on NIS client.\n"),
src/chfn.c:690:			         ("%s: '%s' is the NIS master for this client.\n"),
src/chsh.c:534:		         _("%s: cannot change user '%s' on NIS client.\n"),
src/chsh.c:540:			         _("%s: '%s' is the NIS master for this client.\n"),
src/groupmod.c:842:	 * Now make sure it isn't an NIS group.
src/groupmod.c:849:		         _("%s: group %s is a NIS group\n"),
src/groupmod.c:855:			         _("%s: %s is the NIS master\n"),
src/groupdel.c:438:	 * Make sure this isn't a NIS group
src/groupdel.c:445:		         _("%s: group '%s' is a NIS group\n"),
src/groupdel.c:451:			         _("%s: %s is the NIS master\n"),
src/userdel.c:1135:	 * Now make sure it isn't an NIS user.
src/userdel.c:1142:		         _("%s: user %s is a NIS user\n"), Prog, user_name);
src/userdel.c:1146:			         _("%s: %s is the NIS master\n"),
src/usermod.c:269:		 * Don't add this group if they are an NIS group. Tell the
src/usermod.c:274:			         _("%s: group '%s' is a NIS group.\n"),
src/usermod.c:1278:	 * Now make sure it isn't an NIS user.
src/usermod.c:1285:		         _("%s: user %s is a NIS user\n"),
src/usermod.c:1291:			         _("%s: %s is the NIS master\n"),
src/pwck.c:380:		 * If this is a NIS line, skip it. You can't "know" what NIS
src/pwck.c:381:		 * is going to do without directly asking NIS ...
src/pwck.c:701:		 * If this is a NIS line, skip it. You can't "know" what NIS
src/pwck.c:702:		 * is going to do without directly asking NIS ...
src/newgrp.c:630:	 * For split groups (due to limitations of NIS), check all
src/passwd.c:816:				/* -r repository (files|nis|nisplus) */
src/grpck.c:468:		 * Skip all NIS entries.
src/useradd.c:823:		 * Don't add this group if they are an NIS group. Tell
src/useradd.c:828:			         _("%s: group '%s' is a NIS group.\n"),
```

```c
$ grep -rnI 'NIS'  | grep -v -e ^po -e ^NEWS -e ^man -e ^ChangeLog -e ^tests
doc/HOWTO:227:     NFS (Network File System), NIS, or some other method.
doc/HOWTO:882:  If you insist on the standard one, or you need NIS support, Sunsite
lib/prefix_flag.c:121:#ifdef	USE_NIS
lib/prefix_flag.c:122:		__setspNIS(0); /* disable NIS for now, at least until it is properly supporting a "prefix" */
lib/commonio.c:528: * New entries are inserted before the first NIS entry.  Order is preserved
lib/commonio.c:531:#ifndef KEEP_NIS_AT_END
lib/commonio.c:532:#define KEEP_NIS_AT_END 1
lib/commonio.c:535:#if KEEP_NIS_AT_END
lib/commonio.c:540: * Insert an entry between the regular entries, and the NIS entries.
lib/commonio.c:568:#endif				/* KEEP_NIS_AT_END */
lib/commonio.c:736:#if KEEP_NIS_AT_END
lib/commonio.c:742:#if KEEP_NIS_AT_END
lib/commonio.c:751:#if KEEP_NIS_AT_END
lib/commonio.c:768:#if KEEP_NIS_AT_END
lib/commonio.c:785:#if KEEP_NIS_AT_END
lib/commonio.c:794:#if KEEP_NIS_AT_END
lib/commonio.c:1096:#if KEEP_NIS_AT_END
lib/commonio.c:1098:#else				/* !KEEP_NIS_AT_END */
lib/commonio.c:1100:#endif				/* !KEEP_NIS_AT_END */
lib/shadow.c:26:#ifdef	USE_NIS
lib/shadow.c:37:#define	IS_NISCHAR(c) ((c)=='+')
lib/shadow.c:45:#ifdef	USE_NIS
lib/shadow.c:48: * __setspNIS - turn on or off NIS searches
lib/shadow.c:51:void __setspNIS (bool flag)
lib/shadow.c:61: * bind_nis - bind to NIS server
lib/shadow.c:87:#ifdef	USE_NIS
lib/shadow.c:154:	 * to turn on NIS processing.
lib/shadow.c:158:#ifdef	USE_NIS
lib/shadow.c:159:	if (IS_NISCHAR (fields[0][0])) {
lib/shadow.c:168:	 * incorrectly formatted number, unless we are using NIS.
lib/shadow.c:174:#ifdef	USE_NIS
lib/shadow.c:191:#ifdef	USE_NIS
lib/shadow.c:208:#ifdef	USE_NIS
lib/shadow.c:239:#ifdef	USE_NIS
lib/shadow.c:257:#ifdef	USE_NIS
lib/shadow.c:275:#ifdef	USE_NIS
lib/shadow.c:293:#ifdef	USE_NIS
lib/shadow.c:319:#ifdef	USE_NIS
lib/shadow.c:329:#ifdef	USE_NIS
lib/shadow.c:330:		if (nis_ignore && IS_NISCHAR (buf[0])) {
lib/shadow.c:345:#ifdef	USE_NIS
lib/shadow.c:353:#ifdef	USE_NIS
lib/shadow.c:371:		 * If this entry began with a NIS escape character, we have
lib/shadow.c:376:		if (IS_NISCHAR (val->sp_namp[0])) {
lib/shadow.c:384:		 * If this isn't a NIS user and this isn't an escape to go
lib/shadow.c:385:		 * use a NIS map, it must be a regular local user.
lib/shadow.c:392:		 * If this is an escape to use an NIS map, switch over to
lib/shadow.c:441:#ifdef	USE_NIS
lib/shadow.c:448:#ifdef	USE_NIS
lib/shadow.c:482:#ifdef	USE_NIS
lib/shadow.c:485:	 * other three files.  I can't just blindly turn off NIS because
lib/shadow.c:487:	 * that case, I never discover that NIS is present.
lib/shadow.c:500:#ifdef	USE_NIS
lib/gshadow.c:33:#ifdef	USE_NIS
lib/gshadow.c:44:#define	IS_NISCHAR(c) ((c)=='+')
lib/gshadow.c:47:#ifdef	USE_NIS
lib/gshadow.c:49: * bind_nis - bind to NIS server
lib/gshadow.c:93:#ifdef	USE_NIS
lib/gshadow.c:157:#ifdef	USE_NIS
lib/gshadow.c:158:		if (!IS_NISCHAR (fields[0][0])) {
lib/gshadow.c:212:#ifdef	USE_NIS
lib/gshadow.c:240:#ifdef	USE_NIS
lib/gshadow.c:241:		if (nis_ignore && IS_NISCHAR (buf[0])) {
lib/gshadow.c:256:#ifdef	USE_NIS
lib/gshadow.c:264:#ifdef	USE_NIS
lib/gshadow.c:283:		 * If this entry began with a NIS escape character, we have
lib/gshadow.c:288:		if (IS_NISCHAR (val->sg_name[0])) {
lib/gshadow.c:297:		 * If this isn't a NIS group and this isn't an escape to go
lib/gshadow.c:298:		 * use a NIS map, it must be a regular local group.
lib/gshadow.c:306:		 * If this is an escape to use an NIS map, switch over to
lib/gshadow.c:356:#ifdef	USE_NIS
lib/gshadow.c:363:#ifdef	USE_NIS
lib/gshadow.c:399:#ifdef	USE_NIS
lib/gshadow.c:410:#ifdef	USE_NIS
etc/login.access:30:# If you run NIS you can use @netgroupname in host or user patterns; this
src/chfn.c:674:#ifdef	USE_NIS
src/chfn.c:678:	if (__ispwNIS ()) {
src/chfn.c:683:		         _("%s: cannot change user '%s' on NIS client.\n"),
src/chfn.c:690:			         ("%s: '%s' is the NIS master for this client.\n"),
src/chsh.c:525:#ifdef	USE_NIS
src/chsh.c:529:	if (__ispwNIS ()) {
src/chsh.c:534:		         _("%s: cannot change user '%s' on NIS client.\n"),
src/chsh.c:540:			         _("%s: '%s' is the NIS master for this client.\n"),
src/groupmod.c:840:#ifdef	USE_NIS
src/groupmod.c:842:	 * Now make sure it isn't an NIS group.
src/groupmod.c:844:	if (__isgrNIS ()) {
src/groupmod.c:849:		         _("%s: group %s is a NIS group\n"),
src/groupmod.c:855:			         _("%s: %s is the NIS master\n"),
src/groupdel.c:436:#ifdef	USE_NIS
src/groupdel.c:438:	 * Make sure this isn't a NIS group
src/groupdel.c:440:	if (__isgrNIS ()) {
src/groupdel.c:445:		         _("%s: group '%s' is a NIS group\n"),
src/groupdel.c:451:			         _("%s: %s is the NIS master\n"),
src/userdel.c:1132:#ifdef	USE_NIS
src/userdel.c:1135:	 * Now make sure it isn't an NIS user.
src/userdel.c:1137:	if (__ispwNIS ()) {
src/userdel.c:1142:		         _("%s: user %s is a NIS user\n"), Prog, user_name);
src/userdel.c:1146:			         _("%s: %s is the NIS master\n"),
src/userdel.c:1151:#endif				/* USE_NIS */
src/usermod.c:267:#ifdef	USE_NIS
src/usermod.c:269:		 * Don't add this group if they are an NIS group. Tell the
src/usermod.c:272:		if (__isgrNIS ()) {
src/usermod.c:274:			         _("%s: group '%s' is a NIS group.\n"),
src/usermod.c:1276:#ifdef	USE_NIS
src/usermod.c:1278:	 * Now make sure it isn't an NIS user.
src/usermod.c:1280:	if (__ispwNIS ()) {
src/usermod.c:1285:		         _("%s: user %s is a NIS user\n"),
src/usermod.c:1291:			         _("%s: %s is the NIS master\n"),
src/pwck.c:380:		 * If this is a NIS line, skip it. You can't "know" what NIS
src/pwck.c:381:		 * is going to do without directly asking NIS ...
src/pwck.c:701:		 * If this is a NIS line, skip it. You can't "know" what NIS
src/pwck.c:702:		 * is going to do without directly asking NIS ...
src/newgrp.c:630:	 * For split groups (due to limitations of NIS), check all
src/grpck.c:468:		 * Skip all NIS entries.
src/useradd.c:821:#ifdef	USE_NIS
src/useradd.c:823:		 * Don't add this group if they are an NIS group. Tell
src/useradd.c:826:		if (__isgrNIS ()) {
src/useradd.c:828:			         _("%s: group '%s' is a NIS group.\n"),
alx@debian:~/src/shadow/shadow/str2i$ grep -rnI 'NIS'  | grep -v -e ^po -e ^NEWS -e ^man -e ^ChangeLog -e ^tests
```

Should I pull the trigger?  :-)

**Comment by jubalh** [comment]
It seems to be there since the import from svn to git. But I also couldn't find anywhere how to enable it.
Some commits add new code with the USE_NIS case but I wonder if they tested those code paths or just made it align with other places that were similar.

**Comment by alejandro-colomar** [comment]
> It seems to be there since the import from svn to git. But I also couldn't find anywhere how to enable it. Some commits add new code with the USE_NIS case but I wonder if they tested those code paths or just made it align with other places that were similar.

A funny story about a similar thing:  In nginx unit, where I worked in the past, there is a flag to build the program without support for Unix sockets (they said it's for optimizing).  Everybody blindly added code for it.  Then, one day I added a feature that touched that code, and tested with and without the flag, to test my changes.  Guess what?  It didn't even build, due to a syntax error (a closing brace was in a wrong place).  And the syntax error had been introduced 5 years ago, so nobody had tested that flag in the previous 5 years (the program isn't much older than that; that's like saying the flag was tested when it was written --hopefully-- and never tested again).  Then I suggested just removing that flag and all the dead code it had.  The suggestion was rejected.  There are still no tests for that flag.
